notifiers/matrix.go

```diff
@@ -0,0 +1,181 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package notifiers
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/hunterlong/statping/core/notifier"
+	"github.com/hunterlong/statping/types"
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+type matrix struct {
+	*notifier.Notification
+}
+
+var Matrix = &matrix{&notifier.Notification{
+	Method:      "matrix",
+	Title:       "Matrix",
+	Description: "Receive notifications on your Matrix room when a service has an issue. You must get a username and token. It is recommended to create a dedicated user for this.",
+	Author:      "Tero Vierimaa",
+	AuthorUrl:   "https://github.com/tryffel",
+	Icon:        "far fa-bell",
+	Delay:       5 * time.Second,
+	Form: []notifier.NotificationForm{{
+		Type:        "text",
+		Title:       "Homeserver url",
+		Placeholder: "https://matrix.org",
+		SmallText:   "Enter homeserver url, http(s) included",
+		DbField:     "host",
+		Required:    true,
+	}, {
+		Type:        "text",
+		Title:       "Room Id",
+		Placeholder: "!MzCSpOkYukLxYisAbY:matrix.org",
+		SmallText:   "Enter full room id",
+		DbField:     "var1",
+		Required:    true,
+	}, {
+		Type:        "text",
+		Title:       "Token",
+		Placeholder: "MDAxyz...",
+		SmallText:   "Enter user token",
+		DbField:     "api_secret",
+		Required:    true,
+	}, {
+		Type:        "text",
+		Title:       "Event type",
+		Placeholder: "m.text",
+		SmallText:   "Either m.text (normal message) or m.notice (silent message)",
+		DbField:     "var2",
+		Required:    false,
+	}},
+}}
+
+// matrixUrl, params: room, nonce
+var matrixUrlPath = "/_matrix/client/r0/rooms/%s/send/m.room.message/%d"
+
+func (m *matrix) Select() *notifier.Notification {
+	return m.Notification
+}
+
+func (m *matrix) Send(msg interface{}) error {
+	message := msg.(string)
+	return m.sendMsg(message, m.Var2)
+}
+
+func (m *matrix) OnFailure(s *types.Service, f *types.Failure) {
+	msg := fmt.Sprintf("Your service '%v' is currently offline!", s.Name)
+	m.AddQueue(fmt.Sprintf("service_%v", s.Id), msg)
+}
+
+func (m *matrix) OnSuccess(s *types.Service) {
+	if !s.Online || !s.SuccessNotified {
+		m.ResetUniqueQueue(fmt.Sprintf("service_%v", s.Id))
+		var msg interface{}
+		if s.UpdateNotify {
+			s.UpdateNotify = false
+		}
+		msg = s.DownText
+		m.AddQueue(fmt.Sprintf("service_%v", s.Id), msg)
+	}
+}
+
+func (m *matrix) OnSave() error {
+	// Validate event type
+	if m.Var2 != "m.text" && m.Var2 != "m.notice" {
+		return fmt.Errorf("only 'm.text' or 'm.notice' supported")
+	}
+	return nil
+}
+
+func (m *matrix) OnTest() error {
+	return m.sendMsg("Testing Matrix Notifier on you Statping server", m.Var2)
+}
+
+func (m *matrix) sendMsg(msg, msgType string) error {
+	if msgType == "" {
+		msgType = "m.text"
+	}
+
+	data := map[string]string{
+		"msgtype": msgType,
+		"body":    msg,
+	}
+
+	nonce := rand.Intn(10000)
+	url := m.Host + matrixUrlPath
+	url = fmt.Sprintf(m.Host+matrixUrlPath, m.Var1, nonce)
```

Comment:
Can you please explain for what this `nonce` is and why you only use a "random" Integer – which is _not_ really random since you are using the `math/rand` package and not `crypto/rand` – which generates a number between _0_ and _10000_

Reply:
Also, it would be nice if you could explain what  `m.Var1` is used for in your Notifier.

Reply:
Okay I looked at Matrix spec a bit closer and implemented proper transaction Id (nonce in pervious commit). It's only requirement is to be unique per client through time (with same access token). 

Var1: basically let the user choose whether message is displayed as a user message or as a bot message. Advanced configuration but IMO useful, since bot messages are muted by default and that doesn't fit the purpose of alarms.

Reply:
Ok, thanks for clarifying @tryffel 😊